File: scripts/sap_connector.py
import requests
import json
import logging
from config import SERVICE_LAYER_URL, COMPANY_DB, USERNAME, PASSWORD, CIPHER

# Importaciones para el adaptador SSL/TLS
from requests.adapters import HTTPAdapter
from requests.packages.urllib3.util.ssl_ import create_urllib3_context

# Configurar logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

# --- Adaptador Personalizado ---
# Bloque necesario para que la librería requests/urllib3 reconozca el cipher obsoleto
try:
    requests.packages.urllib3.util.ssl_.DEFAULT_CIPHERS += ":" + CIPHER
except AttributeError:
    pass 

class LegacyTLSAdapter(HTTPAdapter):
    """
    Adaptador personalizado que fuerza el uso de TLSv1 (TLS 1.0) y la cipher suite AES256-SHA.
    Deshabilita la verificación del hostname para resolver el error 'ValueError' cuando se usa verify=False.
    """
    
    def init_poolmanager(self, connections, maxsize, block=False):
        # Crear un contexto SSL/TLS específico
        ctx = create_urllib3_context()
        
        # --- FIX CRUCIAL: Desactivar la verificación de hostname ---
        # Resuelve el ValueError al combinar verify=False con la verificación del host.
        ctx.check_hostname = False
        
        try:
            ctx.minimum_version = requests.packages.urllib3.util.ssl_.TLSVersion.TLSv1_0 
            ctx.set_ciphers(CIPHER)
        except AttributeError:
            pass 
        
        self.poolmanager = requests.packages.urllib3.poolmanager.PoolManager(
            num_pools=connections,
            maxsize=maxsize,
            block=block,
            ssl_context=ctx
        )

def get_sap_session():
    """Realiza el login con SAP B1 y retorna la sesión de requests (con la cookie B1SESSION)."""
    
    login_url = f"{SERVICE_LAYER_URL}/Login"
    login_payload = {
        "CompanyDB": COMPANY_DB,
        "UserName": USERNAME,
        "Password": PASSWORD
    }
    
    # Crea la sesión e instala el adaptador
    session = requests.Session()
    session.mount("https://", LegacyTLSAdapter())
    
    # Configuración de headers e Ignorar certificado (equivale a --insecure)
    session.verify = False 
    session.headers.update({'Content-Type': 'application/json'})

    try:
        response = session.post(login_url, data=json.dumps(login_payload))
        response.raise_for_status()
        
        logging.info("Login exitoso. Sesión establecida.")
        return session
        
    except requests.exceptions.RequestException as e:
        logging.error(f"Fallo en el login: {e}")
        if response is not None:
            logging.error(f"Respuesta del servidor: {response.status_code}, {response.text}")
        return None

def sap_logout(session):
    """Cierra la sesión de SAP B1."""
    if session:
        try:
            session.post(f"{SERVICE_LAYER_URL}/Logout")
            logging.info("Sesión cerrada correctamente.")
        except requests.exceptions.RequestException as e:
            logging.warning(f"Error al cerrar sesión, pero continuando: {e}")

File: scripts/sapConnector.js
import axios from 'axios';
import https from 'https';
import dotenv from 'dotenv';

// Cargar variables de entorno del archivo .env
dotenv.config();

// Obtener variables de entorno
const SERVICE_LAYER_URL = process.env.SAP_SERVICE_LAYER_URL;
const COMPANY_DB = process.env.SAP_COMPANY_DB;
const USERNAME = process.env.SAP_USERNAME;
const PASSWORD = process.env.SAP_PASSWORD;
const SECMETHOD = process.env.SEC_METHOD;

const agent = new https.Agent({
    rejectUnauthorized: false,
    secureProtocol: SECMETHOD,
});

/**
 * Realiza el login al Service Layer y retorna un objeto de sesión Axios configurado.
 * @returns {AxiosInstance | null} Objeto Axios con la cookie B1SESSION inyectada.
 */
export async function getSapSession() {
    const loginUrl = `${SERVICE_LAYER_URL}/Login`;
    const payload = {
        CompanyDB: COMPANY_DB,
        UserName: USERNAME,
        Password: PASSWORD,
    };

    try {
        const response = await axios.post(loginUrl, payload, {
            httpsAgent: agent,
            headers: { 'Content-Type': 'application/json' },
            // Permite a Axios manejar los redirects y cookies, aunque se recomienda inyección manual
            maxRedirects: 0, 
        });

        // El Service Layer devuelve la cookie en el header 'set-cookie'
        const sessionCookie = response.headers['set-cookie'];
        
        if (!sessionCookie) {
            throw new Error('Login failed: No session cookie received.');
        }

        // Crear una nueva instancia de Axios con la cookie inyectada para reutilización
        const sapAxiosInstance = axios.create({
            baseURL: SERVICE_LAYER_URL,
            httpsAgent: agent,
            headers: {
                'Cookie': sessionCookie.join('; '), // Unir todas las cookies (B1SESSION y ROUTEID)
                'Content-Type': 'application/json'
            },
            // Asegurar que el certificado se ignore en todas las llamadas
            validateStatus: status => status >= 200 && status < 300,
        });

        console.log('✅ Login exitoso. Sesión establecida.');
        return sapAxiosInstance;

    } catch (error) {
        console.error('❌ Fallo al conectar con SAP B1:');
        if (error.response) {
            console.error(`Estado: ${error.response.status}`);
            console.error('Cuerpo del error:', error.response.data);
        } else if (error.code === 'DEPTH_ZERO_SELF_SIGNED_CERT' || error.message.includes('SSL')) {
            console.error('Error SSL: Verifique si el servidor está usando TLS 1.0 o un certificado autofirmado.');
        } else {
            console.error(error.message);
        }
        return null;
    }
}

/**
 * Cierra la sesión activa de SAP B1.
 * @param {AxiosInstance} sessionInstance 
 */
export async function sapLogout(sessionInstance) {
    if (sessionInstance) {
        try {
            await sessionInstance.post('/Logout');
            console.log('🚪 Sesión de SAP B1 cerrada.');
        } catch (error) {
            console.warn('Advertencia: Error al cerrar la sesión, pero el proceso continúa.');
        }
    }
}